File: python/string/sortString.py
#!/usr/bin/env python3
#coding=utf-8

''' leetcode issue 1370 '''

class Solution:
    def sortString(self, s: str) -> str:
        sup = [0] * 26
        for i in s:
            sup[ord(i) - ord('a')] += 1

        ret_len = 0
        ret = ''
        n = len(s)
        while ret_len < n:
            for i in range(26):
                if(sup[i]):
                    ret += chr(i + ord('a'))
                    ret_len += 1
                    sup[i]-=1
            for i in range(25,-1,-1):
                if(sup[i]):
                    ret += chr(i + ord('a'))
                    ret_len += 1
                    sup[i]-=1
        return ret

File: c++/string/sortString.cc
#include <string>
#include <vector>

using namespace std;

/* leetcode issue 1370 */
class Solution {
public:
    string sortString(string s) {
        vector<int> sup(26, 0);
        for(char c : s) {
            sup[c-'a']++;
        }

        string ret;
        while(ret.size() < s.size()) {
            for(size_t i = 0; i < sup.size(); ++i) {
                if(sup[i] != 0) {
                    ret.push_back(i+'a');
                    --sup[i];
                }
            }
            /* can't use size_t here */
            for(int i = 25; i >= 0; --i) {
                if(sup[i] != 0) {
                    ret.push_back(i+'a');
                    --sup[i];
                }
            }
        }

        return ret;
    }
};

